pkg/protocol/engine/engine.go

```diff
@@ -216,6 +216,41 @@ func New(
 	)
 }
 
+func (e *Engine) ProcessBlockFromPeer(block *model.Block, source peer.ID) {
+	e.Filter.ProcessReceivedBlock(block, source)
+	e.Events.BlockProcessed.Trigger(block.ID())
+}
+
+// Reset resets the component to a clean state as if it was created at the last commitment.
+func (e *Engine) Reset() {
+	e.Workers.WaitChildren()
```

Comment:
Because of the lack of locking on all the `Reset` function you are probably assuming that no task will be submitted to any of these components after this `e.Workers.WaitChildren()`.
How are we 100% of that?

Reply:
You have to make sure that access to the engine is blocked from the outside, inside you don't have any guarantees, which is something we anyway need to ensure to not accidentaly feed old blocks to the engine after the reset.

Putting a mutex there would make it thread safe for concurrent use but it would not get rid of this requirement, so I think it might be better to put that mental burden on the dev to make sure its correct rather than assuming correctness through the presence of a mutex and making the dev ignore this requirement.